src/number_theory/modular_forms/basic.lean

```diff
@@ -0,0 +1,302 @@
+
+instance has_zero : has_zero (modular_form Γ k) :=
+⟨{ to_fun := 0,
+  slash_action_eq' := slash_action.zero_slash _,
+  hol' := (λ _, mdifferentiable_at_const 𝓘(ℂ, ℂ) 𝓘(ℂ, ℂ)),
+  bdd_at_infty' := λ A, by simpa using zero_form_is_bounded_at_im_infty }⟩
+
+@[simp] lemma coe_zero : ⇑(0 : modular_form Γ k) = (0 : ℍ → ℂ) := rfl
+
+@[simp] lemma zero_apply (z : ℍ) : (0 : modular_form Γ k) z = 0 := rfl
+
+section
+variables {α : Type*} [has_smul α ℂ] [is_scalar_tower α ℂ ℂ]
+
+instance has_smul : has_smul α (modular_form Γ k) :=
+⟨ λ c f,
+  { to_fun := c • f,
+    hol' := by simpa using f.hol'.const_smul (c • (1 : ℂ)),
+    bdd_at_infty' := λ A, by simpa using (f.bdd_at_infty' A).const_smul_left (c • (1 : ℂ)),
+     .. c • (f : slash_invariant_form Γ k)}⟩
+
+@[simp] lemma coe_smul (f : (modular_form Γ k)) (n : α) : ⇑(n • f) = n • f := rfl
+@[simp] lemma smul_apply (f : (modular_form Γ k)) (n : α) (z : ℍ) :
+   (n • f) z = n • (f z) := rfl
+end
+
+instance has_neg : has_neg (modular_form Γ k) :=
+⟨λ f,
+  { to_fun := -f,
+    hol' := f.hol'.neg,
+    bdd_at_infty':= λ A, by simpa using (f.bdd_at_infty' A).neg,
+    .. -(f : slash_invariant_form Γ k) }⟩
+
+@[simp] lemma coe_neg (f : modular_form Γ k) : ⇑(-f) = -f := rfl
+
+@[simp] lemma neg_apply (f : modular_form Γ k) (z : ℍ) : (-f) z = - (f z) := rfl
+
+instance has_sub : has_sub (modular_form Γ k) :=
+⟨ λ f g, f + -g ⟩
+
+@[simp] lemma coe_sub (f g : (modular_form Γ k)) : ⇑(f - g) = f - g := rfl
+
+@[simp] lemma sub_apply (f g : modular_form Γ k) (z : ℍ) : (f - g) z = f z - g z := rfl
+
+instance : add_comm_group (modular_form Γ k) :=
+fun_like.coe_injective.add_comm_group _ rfl coe_add coe_neg coe_sub coe_smul coe_smul
+
+/--Additive coercion from `modular_form` to `ℍ → ℂ`. -/
+def coe_hom : (modular_form Γ k) →+ (ℍ → ℂ) :=
+{ to_fun := λ f, f,
+  map_zero' := coe_zero,
+  map_add' := λ _ _, rfl }
+
+instance : module ℂ (modular_form Γ k) :=
+function.injective.module ℂ (coe_hom) fun_like.coe_injective (λ _ _, rfl)
+
+instance : inhabited (modular_form Γ k) := ⟨0⟩
+
+/--The modular form of weight `k_1 + k_2` given by the product of two modular forms of weights
+`k_1` and `k_2`. -/
+def mul {k_1 k_2 : ℤ} {Γ : subgroup SL(2, ℤ)} (f : (modular_form Γ k_1))
+  (g : (modular_form Γ k_2)) : (modular_form Γ (k_1 + k_2)) :=
+{ to_fun := f * g,
+  slash_action_eq' := λ A, by simp_rw [mul_slash_subgroup, modular_form_class.slash_action_eq],
+  hol' := f.hol'.mul g.hol',
+  bdd_at_infty' := λ A, by simpa using (f.bdd_at_infty' A).mul (g.bdd_at_infty' A) }
+
+@[simp] lemma mul_coe {k_1 k_2 : ℤ} {Γ : subgroup SL(2, ℤ)} (f : (modular_form Γ k_1))
+  (g : (modular_form Γ k_2)) : ((f.mul g) : ℍ → ℂ) = f * g := rfl
+
+instance : has_one (modular_form Γ 0) :=
+⟨{  hol' := λ x, mdifferentiable_at_const 𝓘(ℂ, ℂ) 𝓘(ℂ, ℂ),
+    bdd_at_infty' := λ A, (congr_arg is_bounded_at_im_infty (is_invariant_one A)).mpr
+    (filter.const_bounded_at_filter _ 1),
```

Comment:
Obviously we want this proof to be
```suggestion
    bdd_at_infty' := λ A, by simpa using at_im_infty.const_bounded_at_filter (1:ℂ),
```
I played with it a bit; I think you need to add 
```lean
@[simp] lemma one_coe_eq_one : ((1 : slash_invariant_form Γ 0) : ℍ → ℂ) = 1 := rfl
```
as a simp-lemma after the `has_one` instance for `slash_invariant_form`.